### DISCUSSION TITLE
java.lang.NoSuchMethodError: 'javax.ws.rs.core.Response org.keycloak.forms.login.LoginFormsProvider.createForm'

### DISCUSSION THREAD

**Discussion by vietanh2909** [dasniko/keycloak-2fa-sms-authenticator#40] [Q&A]
Hi! I have error after login:

Uncaught server error: java.lang.NoSuchMethodError: 'javax.ws.rs.core.Response org.keycloak.forms.login.LoginFormsProvider.createForm(java.lang.String)'
        at dasniko.keycloak.authenticator.SmsAuthenticator.authenticate(SmsAuthenticator.java:56)

source code: 
context.challenge(context.form().setAttribute("realm", context.getRealm()).createForm(TPL_CODE));

I try repleace by:
session.getProvider(LoginFormsProvider.class).createForm(TPL_CODE);

But the error still occurs.

![er](https://github.com/dasniko/keycloak-2fa-sms-authenticator/assets/26007180/527917e3-ce34-493b-8982-8953ce960984)

**Comment by vietanh2909** [comment]
My pom

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
				 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
				 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>dasniko</groupId>
	<artifactId>keycloak-2fa-sms-authenticator</artifactId>
	<version>1.0-SNAPSHOT</version>

	<properties>
		<maven.compiler.source>11</maven.compiler.source>
		<maven.compiler.target>11</maven.compiler.target>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
	</properties>

	<dependencies>
		<dependency>
			<groupId>org.keycloak</groupId>
			<artifactId>keycloak-services</artifactId>
		</dependency>
		<dependency>
			<groupId>org.keycloak</groupId>
			<artifactId>keycloak-server-spi</artifactId>
		</dependency>
		<dependency>
			<groupId>org.keycloak</groupId>
			<artifactId>keycloak-server-spi-private</artifactId>
		</dependency>
		
		<dependency>
			<groupId>com.google.code.gson</groupId>
			<artifactId>gson</artifactId>
			<version>2.8.5</version>
		</dependency>
		<dependency>
			<groupId>org.projectlombok</groupId>
			<artifactId>lombok</artifactId>
			<version>1.18.28</version>
			<scope>compile</scope>
		</dependency>

		
		<dependency>
			<groupId>com.google.auto.service</groupId>
			<artifactId>auto-service</artifactId>
			<version>1.0-rc5</version>
			<scope>provided</scope>
			<optional>true</optional>
		</dependency>


	</dependencies>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.keycloak</groupId>
				<artifactId>keycloak-parent</artifactId>
				<version>21.0.0</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<build>
		<finalName>${project.artifactId}-${project.version}</finalName>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-jar-plugin</artifactId>
				<configuration>
					<archive>
						<manifestEntries>
							<Dependencies>com.google.code.gson</Dependencies>
						</manifestEntries>
					</archive>
				</configuration>
			</plugin>
		</plugins>
	</build>

</project>

**Comment by dasniko** [comment]
> Uncaught server error: java.lang.NoSuchMethodError: 'javax.ws.rs.core.Response

Seems that you are using a pre-kc-22 version of this provider in a KC22+ environment.
See release notes: https://www.keycloak.org/2023/07/keycloak-2200-released.html#_upgrade_to_quarkus_3_x

**Reply by vietanh2909** [reply]
Thanks for reply.
Could you make suggestion for me to replace it?

**Reply by vietanh2909** [reply]
Hi @dasniko,
I down version to 20.0.0 the issue has resolve. Thanks
![image](https://github.com/dasniko/keycloak-2fa-sms-authenticator/assets/26007180/e21d0c55-d08f-4213-922b-5de9678576cf)
